bigquery/storage_client.go

```diff
@@ -0,0 +1,159 @@
+	maxStreamCount int
+	maxWorkerCount int
+}
+
+func defaultReadClientSettings() *readClientSettings {
+	maxWorkerCount := runtime.GOMAXPROCS(0)
+	return &readClientSettings{
+		maxStreamCount: 0,
+		maxWorkerCount: maxWorkerCount,
+	}
+}
+
+// newReadClient instantiates a new storage read client.
+func newReadClient(ctx context.Context, projectID string, opts ...option.ClientOption) (c *readClient, err error) {
+	numConns := runtime.GOMAXPROCS(0)
+	if numConns > 4 {
+		numConns = 4
+	}
+	o := []option.ClientOption{
+		option.WithGRPCConnectionPool(numConns),
+		option.WithUserAgent(fmt.Sprintf("%s/%s", userAgentPrefix, internal.Version)),
+	}
+	o = append(o, opts...)
+
+	rawClient, err := storage.NewBigQueryReadClient(ctx, o...)
+	if err != nil {
+		return nil, err
+	}
+	rawClient.SetGoogleClientInfo("gccl", internal.Version)
+
+	// Handle project autodetection.
+	projectID, err = detect.ProjectID(ctx, projectID, "", opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	rc := &readClient{
+		rawClient: rawClient,
+		projectID: projectID,
+		settings:  defaultReadClientSettings(),
+	}
+
+	return rc, nil
+}
+
+// close releases resources held by the client.
+func (c *readClient) close() error {
+	if c.rawClient == nil {
+		return fmt.Errorf("already closed")
+	}
+	c.rawClient.Close()
+	c.rawClient = nil
+	return nil
+}
+
+// sessionForTable establishes a new session to fetch from a table using the Storage API
+func (c *readClient) sessionForTable(ctx context.Context, table *Table) (*readSession, error) {
+	tableID, err := table.Identifier(StorageAPIResourceID)
+	if err != nil {
+		return nil, err
+	}
+
+	rs := &readSession{
+		ctx:                   ctx,
+		table:                 table,
+		tableID:               tableID,
+		settings:              c.settings,
+		readRowsFunc:          c.rawClient.ReadRows,
+		createReadSessionFunc: c.rawClient.CreateReadSession,
+	}
+	return rs, nil
+}
+
+// ReadSession is the abstraction over a storage API read session.
+type readSession struct {
+	settings *readClientSettings
+
+	ctx     context.Context
+	table   *Table
+	tableID string
+
+	bqSession *storagepb.ReadSession
+
+	// decouple from readClient to enable testing
+	createReadSessionFunc func(context.Context, *storagepb.CreateReadSessionRequest, ...gax.CallOption) (*storagepb.ReadSession, error)
+	readRowsFunc          func(context.Context, *storagepb.ReadRowsRequest, ...gax.CallOption) (storagepb.BigQueryRead_ReadRowsClient, error)
+}
+
+// Start initiates a read session
+func (rs *readSession) start() error {
+	createReadSessionRequest := &storagepb.CreateReadSessionRequest{
+		Parent: fmt.Sprintf("projects/%s", rs.table.ProjectID),
+		ReadSession: &storagepb.ReadSession{
+			Table:      rs.tableID,
+			DataFormat: storagepb.DataFormat_ARROW,
+		},
+		MaxStreamCount: int32(rs.settings.maxStreamCount),
+	}
+	rpcOpts := gax.WithGRPCOptions(
+		grpc.MaxCallRecvMsgSize(1024 * 1024 * 129), // TODO: why needs to be of this size
```

Comment:
Default is 4MB: https://github.com/grpc/grpc-go/blob/07ac97c355970d6210cfa5341f2ba0850ec2757b/clientconn.go#L96

Read API can send batches up to 128MB.